wolfcrypt/src/port/Espressif/esp32_sha.c

```diff
@@ -0,0 +1,437 @@
+            ctx->mode = ESP32_SHA_SW;
+            return 0;
+        }
+    }
+    /* check if this sha has been operated as sw or hw, or not yet init */
+    if(ctx->mode == ESP32_SHA_INIT){
+        /* try to lock the hw engine */
+        if(esp_CryptHwMutexLock(&sha_mutex, (TickType_t)0) == 0) {
+            ctx->mode = ESP32_SHA_HW;
+        } else {
+            ESP_LOGI(TAG, "someone used. hw is locked.....");
+            ESP_LOGI(TAG, "the rest of operaion will use sw implementation for this sha");
+            ctx->mode = ESP32_SHA_SW;
+            return 0;
+        }
+    } else {
+        /* this should not happens */
+        ESP_LOGE(TAG, "unexpected error in esp_sha_try_hw_lock.");
+        return -1;
+    }
+#endif
+   /* Enable SHA hardware */
+    periph_module_enable(PERIPH_SHA_MODULE);
+
+    ESP_LOGV(TAG, "leave esp_sha_hw_lock");
+    return ret;
+}
+/*
+* release hw engine
+*/
+void esp_sha_hw_Release( void )
+{
+    ESP_LOGV(TAG, "enter esp_sha_hw_Release");
+
+    /* Disable AES hardware */
+    periph_module_disable(PERIPH_SHA_MODULE);
+#if defined(SINGLE_THREADED)
+    InUse = 0;
+#else
+    /* unlock hw engine for next use */
+    esp_CryptHwMutexUnLock(&sha_mutex);
+#endif
+    ESP_LOGV(TAG, "leave esp_sha_hw_Release");
+}
+/*
+* start sha process by using hw engine
+*/
+void esp_sha_start_process(WC_ESP32SHA* sha, word32 address)
+{
+    ESP_LOGV(TAG, "enter esp_sha_start_process");
+
+    if(sha->isfirstblock){
+        /* start first message block */
+        DPORT_REG_WRITE(address, 1);
+        sha->isfirstblock = 0;
+    } else {
+        /* CONTINU_REG */
+        DPORT_REG_WRITE(address + 0x04, 1);
+   }
+
+   ESP_LOGV(TAG, "leave esp_sha_start_process");
+}
+/*
+* process message block
+*/
+void esp_process_block(WC_ESP32SHA* ctx,  word32 address,
+                                         const word32* data, word32 len)
+{
+    ESP_LOGV(TAG, "enter esp_process_block");
+
+    int i;
+
+    /* check if there are any busy engine */
+    esp_wait_until_idle();
+    /* load message data into hw */
+    for(i=0;i<((len)/(sizeof(word32)));++i){
+        DPORT_REG_WRITE(SHA_TEXT_BASE+(i*4),*(data+i));
+    }
+    /* notify hw to start process */
+    esp_sha_start_process(ctx, address);
+
+    ESP_LOGV(TAG, "leave esp_process_block");
+}
+/*
+* retreive sha digest from memory
+*/
+void esp_digest_state(WC_ESP32SHA* ctx, byte* hash, enum SHA_TYPE sha_type)
+{
+    /* registers */
+    word32 SHA_LOAD_REG = SHA_1_LOAD_REG;
+    word32 SHA_BUSY_REG = SHA_1_BUSY_REG;
+
+    ESP_LOGV(TAG, "enter esp_digest_state");
+
+    /* sanity check */
+    if(sha_type == SHA_INVALID) {
+        ESP_LOGE(TAG, "unexpected error. sha_type is invalid.");
+        return;
+    }
+    
+    SHA_LOAD_REG += (sha_type << 4);
+    SHA_BUSY_REG += (sha_type << 4);
+
+    if(ctx->isfirstblock == 1){
+        /* no hardware use yet. Nothing to do yet */
+        return ;
+    }
+
+    {
```

Comment:
Extra braces here to line 278, that are not required.

Reply:
Removed the braces.